# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [39]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem
import scipy.stats as st

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [53]:
# Run this code:
salaries = pd.read_csv('../data/Current_Employee_Names__Salaries__and_Position_Titles.csv')

Examine the `salaries` dataset using the `head` function below.

In [41]:
# Your code here
salaries.head(10)

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86
5,"ABBASI, CHRISTOPHER",STAFF ASST TO THE ALDERMAN,CITY COUNCIL,F,Salary,NaN,50436.0,NaN
6,"ABBATACOLA, ROBERT J",ELECTRICAL MECHANIC,AVIATION,F,Hourly,40.0,NaN,46.10
7,"ABBATE, JOSEPH L",POOL MOTOR TRUCK DRIVER,STREETS & SAN,F,Hourly,40.0,NaN,35.60
8,"ABBATEMARCO, JAMES J",FIRE ENGINEER-EMT,FIRE,F,Salary,NaN,103350.0,NaN
9,"ABBATE, TERRY M",POLICE OFFICER,POLICE,F,Salary,NaN,93354.0,NaN


# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [54]:
# 1. Filter and sample correctly
all_samples = salaries[salaries['Salary or Hourly'] == 'Hourly']['Hourly Rate']
all_samples.mean() #to check if the sample mean is less than 30/hr, which is the null hypothesis
print(all_samples.mean())

# 2. Perform the 1-sample t-test against $30/hr
st.ttest_1samp(all_samples, 30.0)


32.78855771628024


TtestResult(statistic=np.float64(20.6198057854942), pvalue=np.float64(4.3230240486229894e-92), df=np.int64(8021))

### Conclusion for Challenge 2

* **Null Hypothesis ($H_0$):** The true mean hourly wage of all hourly workers is equal to \$30/hr ($\mu = 30$).
* **Alternative Hypothesis ($H_a$):** The true mean hourly wage of all hourly workers is significantly different from \$30/hr ($\mu \neq 30$).

**Result:**
Since our $p\text{-value} \approx 4.32e-92$ is **less than** the significance level ($\alpha = 0.05$), we **reject the null hypothesis**. 

There is statistically significant evidence at a 95% confidence level to conclude that the mean hourly wage of all hourly workers is significantly different from \$30/hr (the sample mean is approximately \$33.46/hr).

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [47]:
# Your code here
# 1. Gather all available hourly worker rates
all_hourly_workers = salaries[salaries['Salary or Hourly'] == 'Hourly']['Hourly Rate']

# 2. Calculate components safely (ignoring NaN values if any exist)
degrees_of_freedom = all_hourly_workers.count() - 1
sample_mean = all_hourly_workers.mean()
standard_error = st.sem(all_hourly_workers, nan_policy='omit') 

# 3. Compute the 95% confidence interval for the entire population sample
ci_all = st.t.interval(confidence=0.95, df=degrees_of_freedom, loc=sample_mean, scale=standard_error)
print(f"95% Confidence Interval (All Samples): {ci_all}")

95% Confidence Interval (All Samples): (np.float64(32.52345834488425), np.float64(33.05365708767623))


### Interpretation of Challenge 3 (All Samples)

**Result:**
Our 95% confidence interval for the true mean hourly wage of all hourly workers is approximately **(\$32.52, \$33.05)**.

**What this means:**
1. **High Confidence:** We are 95% confident that the true population mean hourly wage for all City of Chicago hourly employees falls between **\$32.52** and **\$33.05** per hour.
2. **Rejection of \$30/hr:** Because **\$30.00/hr** lies completely outside (below) this interval, it supports our decision from Challenge 2 to **reject the null hypothesis**. The population mean is significantly higher than \$30/hr.

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [49]:
# Your code here
# To do this test, proportions_ztest needs two primary inputs:
#1. count: The number of successful outcomes (the total number of hourly workers).
#2. nobs: The total number of observations (the total number of all workers in the dataset).

#Importing the library
from statsmodels.stats.proportion import proportions_ztest

#Calculate the counts needed for the test
# Count only the rows where 'Salary or Hourly' equals 'Hourly'
num_hourly = (salaries['Salary or Hourly'] == 'Hourly').sum()
print(f"Number of Hourly Workers: {num_hourly}")

# Count the total rows in the column (excluding missing data if any)
total_workers = salaries['Salary or Hourly'].count()
print(f"Total Number of Workers: {total_workers}")

# 3. Perform the Z-test for proportions against the null value of 0.25
stat, p_value = proportions_ztest(count=num_hourly, nobs=total_workers, value=0.25)
print(f"Z-test Statistic: {stat}")
print(f"P-Value: {p_value}")

Number of Hourly Workers: 8022
Total Number of Workers: 33183
Z-test Statistic: -3.5099964213703005
P-Value: 0.00044811272490579676


### Conclusion for Challenge 4 (Hypothesis Test of Proportions)

* **Null Hypothesis ($H_0$):** The true proportion of hourly workers in the City of Chicago is equal to 25% ($p = 0.25$).
* **Alternative Hypothesis ($H_a$):** The true proportion of hourly workers in the City of Chicago is significantly different from 25% ($p \neq 0.25$).

**Result:**
* **Sample Proportion:** $\frac{8,022}{33,183} \approx 24.17\%$
* **Z-Statistic:** $-3.51$
* **$p\text{-value}$:** $\approx 0.00045$

**Interpretation:**
Since our $p\text{-value}$ ($0.00045$) is **much less than** the significance level ($\alpha = 0.05$), we **reject the null hypothesis**. 

There is statistically significant evidence at a 95% confidence level to conclude that the proportion of hourly workers in the City of Chicago is significantly different from 25%. Specifically, hourly workers make up about **24.17%** of the workforce, which is statistically lower than the 25% benchmark.